### PR TITLE
Pass communication settings as a k8s secret

### LINKED ISSUE #211: Pass BotKube communication settings as a K8s Secret
**Is your feature request related to a problem? Please describe.**
BotKube config contains sensitive info like Tokens which is passed to backend via ConfigMap. We should pass the communication setting as a K8s Secret. Also, restrict BotKube to access Secret resources using `@BotKube` commands by default

### PR DESCRIPTION
**Pull request by slalwani97**
Fixes #211 

Signed-off-by: Sumit Lalwani <sumit.lalwani97@gmail.com>

### CODE REVIEW

**Comment by PrasadG193** [comment]
Thanks for the PR @slalwani97!
Could you please also update helm chart (https://github.com/infracloudio/botkube/tree/develop/helm/botkube) and all-in-one-tls.yaml (https://github.com/infracloudio/botkube/blob/develop/deploy-all-in-one-tls.yaml)?

You also need split config file we use for e2e test (https://github.com/infracloudio/botkube/tree/develop/test) to make CI build happy

**Comment by slalwani97** [comment]
Sure, I have updated the PR. Please have a look. @PrasadG193

**Comment by PrasadG193** [comment]
LGTM

**Comment by PrasadG193** [comment]
Just a minor suggestion @slalwani97 ,
Can we split config.yaml into - `resource_config.yaml` and `comm_config.yaml`
What do you think?

**Comment by slalwani97** [comment]
Do you mean to change the name of the yaml files, right?
i.e ```config.yaml``` -> ```resource-config.yaml``` and ```communication.yaml``` -> ```comm_config.yaml```

**Comment by PrasadG193** [comment]
@slalwani97 Yes

**Comment by slalwani97** [comment]
Sure @PrasadG193 , I have updated the names with some variable names as well. I have a few doubts: 
1. ```configWatcher``` func in the ```controller.go``` - uses config file, also adds the config file (now resource_config) in the watcher. So is comm_config file should also be added in watcher?
2. ```showControllerConfig``` func in the ```executor.go``` - uses config file too, and reads it. Should the comm_config file be read here as well?
    Probably these lines will break I think: 
    ```
    // hide sensitive info
	c.Communications.Slack.Token = ""
	c.Communications.ElasticSearch.Password = ""
```

**Comment by PrasadG193** [comment]
Thanks @slalwani97 , could you please post the `helm install` command. Just wanted to check if there is any change in that

**Comment by slalwani97** [comment]
@PrasadG193 
```
helm install --version v0.9.1 --name botkube --namespace botkube \
--set communicationConfig.communications.slack.enabled=true \
--set communicationConfig.communications.slack.channel=<SLACK_CHANNEL_NAME> \
--set communicationConfig.communications.slack.token=<SLACK_API_TOKEN_FOR_THE_BOT> \
--set resourceConfig.settings.clustername=<CLUSTER_NAME> \
--set resourceConfig.settings.allowkubectl=<ALLOW_KUBECTL> \
--set image.repository=infracloudio/botkube \
--set image.tag=v0.9.1 \
infracloudio/botkube
```

File: helm/botkube/values.yaml [hunk 1]
```diff
@@ -202,7 +202,19 @@ config:
...
+    # Set true to enable kubectl commands execution
+    allowkubectl: false
+    # Set true to enable config watcher
+    configwatcher: true
+    # Set false to disable upgrade notification
+    upgradeNotifier: true
+
+communicationConfig:
+  # Communication settings
   communications:
```
Comment (PrasadG193) [inline helm/botkube/values.yaml]: `communicationConfig.communications` seems to be redundant. We can remove `communicationConfig` and move `communications` at root level

File: test/comm_config.yaml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    index:
+      name: botkube
+      type: botkube-event
+      shards: 1
+      replicas: 0
+  
+  # Settings for Webhook
+  webhook:
+    enabled: true
+    url: 'WEBHOOK_URL'              # e.g https://example.com:80
```
Comment (PrasadG193) [inline test/comm_config.yaml]: Please add newline.

File: helm/botkube/values.yaml (hunk 1 above)
Comment (slalwani97) [reply helm/botkube/values.yaml]: @PrasadG193 I have updated the PR. PTAL, I am not sure if this is the correct approach for this.

**Comment by slalwani97** [comment]
```
helm install --version v0.9.1 --name botkube --namespace botkube \
--set communications.slack.enabled=true \
--set communications.slack.channel=<SLACK_CHANNEL_NAME> \
--set communications.slack.token=<SLACK_API_TOKEN_FOR_THE_BOT> \
--set config.settings.clustername=<CLUSTER_NAME> \
--set config.settings.allowkubectl=<ALLOW_KUBECTL> \
--set image.repository=infracloudio/botkube \
--set image.tag=v0.9.1 \
infracloudio/botkube
```

**Comment by bhavin192** [comment]
@slalwani97 please add a newline here https://github.com/infracloudio/botkube/pull/226/files#diff-c27d561a78d2228ac0aa26c12fd406a7R16

**Comment by PrasadG193** [comment]
Merged! Thanks for the PR @slalwani97

File: helm/botkube/templates/communicationsecret.yaml [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    helm.sh/chart: {{ include "botkube.chart" . }}
+    app.kubernetes.io/instance: {{ .Release.Name }}
+    app.kubernetes.io/managed-by: {{ .Release.Service }}
+stringData:
+  comm_config.yaml: |
+    # Communication settings
+    communications:
+      {{- with .Values.communications }}
+        {{- toYaml . | nindent 4 }}
+      {{- end }}
```
Comment (PrasadG193) [inline helm/botkube/templates/communicationsecret.yaml]: @slalwani97 Unfortunately we had to revert the commit. The problem is with the indentation here. The secret gets created with following format. That's why it couldn't get unmarshalled into struct 
```
# Communication settings
communications:
elasticsearch:
  enabled: false
  index:
    name: botkube
    replicas: 0
    shards: 1
    type: botkube-event
  password: ELASTICSEARCH_PASSWORD
  server: ELASTICSEARCH_ADDRESS
  username: ELASTICSEARCH_USERNAME
mattermost:
  channel: dev-alerts
  enabled: false
  notiftype: short
  team: MATTERMOST_TEAM
  token: xxxx
  url: xxxx
slack:
  channel: general
  enabled: true
  notiftype: short
  token: xxxxxxxx
webhook:
  enabled: false
  url: WEBHOOK_URL
```

**Comment by PrasadG193** [comment]
Could you please address this and raise another PR?

File: helm/botkube/templates/communicationsecret.yaml (hunk 3 above)
Comment (slalwani97) [reply helm/botkube/templates/communicationsecret.yaml]: Probably changing nindent to 6 will solve the problem.

Comment (slalwani97) [reply helm/botkube/templates/communicationsecret.yaml]:
```
/config $ cat  comm_config.yaml 
# Communication settings
communications:
  elasticsearch:
    enabled: false
    index:
      name: botkube
      replicas: 0
      shards: 1
      type: botkube-event
    password: ELASTICSEARCH_PASSWORD
    server: ELASTICSEARCH_ADDRESS
    username: ELASTICSEARCH_USERNAME
  mattermost:
    channel: MATTERMOST_CHANNEL
    enabled: false
    notiftype: short
    team: MATTERMOST_TEAM
    token: MATTERMOST_TOKEN
    url: MATTERMOST_SERVER_URL
  slack:
    channel: SLACK_CHANNEL
    enabled: false
    notiftype: short
    token: SLACK_API_TOKEN
  webhook:
    enabled: false
    url: WEBHOOK_URL
```

**Comment by slalwani97** [comment]
Addressed in PR #233

**Comment by PrasadG193** [comment]
Thanks @slalwani97

**Pull request merged**